fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/StandardIndexMaintainer.java

```diff
@@ -251,6 +252,100 @@ public <M extends Message> CompletableFuture<Void> update(@Nullable final FDBInd
+                // built, and update the index if so
+                return addedRangeWithKey(oldEntryKey).thenCompose(inRange ->
+                        inRange ? update(oldRecord, newRecord) : AsyncUtil.DONE);
+            } else {
+                // The old and new record use different keys in the source index. Check each one individually,
+                // and then simulate deleting the old record (if needed) and adding the new record (if needed)
+                // Note: index maintainers on the same index are not thread safe, so the index updates need to
+                // be serialized here, but the range set checks can be executed concurrently
+                CompletableFuture<Boolean> oldInRangeFuture = addedRangeWithKey(oldEntryKey);
+                CompletableFuture<Boolean> newInRangeFuture = addedRangeWithKey(newEntryKey);
+                return oldInRangeFuture
+                        .thenCompose(oldInRange -> oldInRange ? update(oldRecord, null) : AsyncUtil.DONE)
+                        .thenCompose(ignore -> newInRangeFuture)
+                        .thenCompose(newInRange -> newInRange ? update(null, newRecord) : AsyncUtil.DONE);
+            }
+        } else {
+            Tuple entryKey = oldEntryKey == null ? newEntryKey : oldEntryKey;
+            if (entryKey == null) {
+                // Both the old and new entries are null. This can happen if there is an index maintenance filter
+                // results in one (or both) of the records being excluded from indexing. If the record(s) is/are
+                // excluded by the filter from the source index, then they must also be excluded from this index,
+                // so it is safe to just do nothing
+                return AsyncUtil.DONE;
+            } else {
+                // One of the entry keys is not null. Check if that one has been built, and then update if it
+                // is in the built range
+                return addedRangeWithKey(entryKey).thenCompose(inRange ->
+                        inRange ? update(oldRecord, newRecord) : AsyncUtil.DONE);
+            }
+        }
+    }
+
+    @Nullable
+    private static <M extends Message> Tuple evaluateSingletonIndexKey(Index index, IndexMaintainer maintainer, @Nullable FDBIndexableRecord<M> record) {
+        if (record == null) {
+            return null;
+        }
+        List<IndexEntry> entries = maintainer.filteredIndexEntries(record);
+        if (entries == null || entries.isEmpty()) {
+            // If there is an IndexMaintenanceFilter on the index, this can return null/an empty list even if the
+            // index key expression always returns a single value. In this case, the record is excluded from the index
+            return null;
+        } else if (entries.size() != 1) {
+            throw new RecordCoreException("index produced incorrect number of entries for use as source index");
```

Comment:
Does this restriction apply for idempotent indexes, or is this a new restriction for non-idempotent indexes?

Reply:
This is not a new restriction: there was already a check in the indexer to make sure the index did not create duplicates. However, it is more important for non-idempotent target indexes.

Along those lines, there are actually two codepaths, a transactional build and a non-transactional build, and it turns out this check was missed in the transactional case. I've added the check in the the other case, which also added a few other missed checks.